### PR TITLE
#164691219 Enable Social login via Google, Facebook and Twitter

### PR DESCRIPTION
**Pull request by reifred**
### What does this PR do?
Enable social users to log into the application using Google, Facebook, and Twitter
### Description of Task to be completed?
* Ensure registered users can log in using their Google, Facebook, and Twitter accounts.
* Enable even unregistered users to register and log in with their Google, Facebook, and Twitter
accounts.
Have the following endpoints working.
* POST `/api/social-auth/google/`
* POST `/api/social-auth/facebook/`
* POST `/api/social-auth/twitter/`
### How should this be manually tested?
* After cloning this repo, cd into it and switch to this branch
* Set the additional environment variables.

`export TWITTER_CONSUMER_KEY='your_twitter_consumer_key'`
`export TWITTER_CONSUMER_SECRET='your_twitter_consumer_secret'`
* Start the development server using `python manage.py runserver`
#### Testing Facebook Social Authentication.
1. Create an application on [facebook developer page](https://developers.facebook.com/).
2. Then get the access_token [here](https://developers.facebook.com/tools/explorer).
3. Test using postman.
POST /api/social-auth/facebook/

Request body
```
{
    "user":{
        "access_token": "your_facebook_access_token"
    }
}
```
Response
```
{
    "user": {
        "email": "reifred33@gmail.com",
        "username": "reifred",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNuIjoicmVpZnJlZCIsImV4cCI6MTU1NzgzNjA4OH0.Q0hgBaJGTAqINl1gPe9BzFO4xtwtm_3QW9yWk3Zw4c8"
    }
}
```
#### Testing Google Social Authentication.
1. Obtain google user_id [here](https://developers.google.com/oauthplayground/).
2. In the scope selection, search for Google OAuth2 API v2 and check all of its three fields.
googlepage
3. Then click the Authorize APIs button
4. Once you have authorized Google+ API to access your information, click Exchange authorization for codes which will give you an access_token and id_token. Copy the id_token and paste it in Postman endpoint body for google login.
token
5. Test using postman.
POST /api/social-auth/google/

Request body
```
{
    "user":{
        "access_token": "your_google_user_id_token"
    }
}
```
Response
```
{
    "user": {
        "email": "reifred33@gmail.com",
        "username": "reifred",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNuIjoicmVpZnJlZCIsImV4cCI6MTU1NzgzNjA4OH0.Q0hgBaJGTAqINl1gPe9BzFO4xtwtm_3QW9yWk3Zw4c8"
    }
}
```
#### Testing Twitter Social Authentication.
1. Sign in [Twitter Developers](https://developer.twitter.com/content/developer-twitter/en.html).Create an App and got to the App details.
2. Navigate to Permissions and authorize access to the email address.
twitter permissions
3. Navigate to Keys and tokens and generate access token and token secret. 
twiiter keys
4. Test using postman.

Request body
```
POST /api/social-auth/twitter/
{
	"user":{
		"access_token": "twitter_access_token",
		"access_token_secret":"twitter_token_secret"
	}
}
```
Response
```
{
    "user": {
        "email": "reifred33@gmail.com",
        "username": "reifred",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNuIjoicmVpZnJlZCIsImV4cCI6MTU1ODU1Mzk1Mn0.a-oLSoFOt_eEfuFb1sHsOyw0O8qUFurTxls2MY-IJjU"
    }
}
```
### What are the relevant pivotal tracker stories?
[#164691219](https://www.pivotaltracker.com/story/show/164691219)

### CODE REVIEW

**Review by RonKbS — CHANGES_REQUESTED** [review]
Great work @reifred.

This implementation is not complete though.
Remember that we are working under the assumption that users that are not developers are interacting with this API.

I would, therefore, recommend that you provide a simple page to redirect users to the social apps.
This is because this redirection is not easily achievable with the third-party clients we're using (postman and insomnia), although I believe it can be done if we had more time.

So, to summarize, users should not interact with the developer tools provided by google, facebook or twitter.

**Comment by reifred** [comment]
> Great work @reifred.
> 
> This implementation is not complete though.
> Remember that we are working under the assumption that users that are not developers are interacting with this API.
> 
> I would, therefore, recommend that you provide a simple page to redirect users to the social apps.
> This is because this redirection is not easily achievable with the third-party clients we're using (postman and insomnia), although I believe it can be done if we had more time.
> 
> So, to summarize, users should not interact with the developer tools provided by google, facebook or twitter.

Thanks @RonKbS for the feedback. Let me research on it.
